botorch/sampling/pathwise/basis/expansions.py

```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/env python3
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+from __future__ import annotations
+
+from typing import Optional
+
+import torch
+from botorch.sampling.pathwise.basis.initializers import GeneralizedLinearInitializer
+from botorch.sampling.pathwise.utils.common import TensorTransform
+from gpytorch.kernels import Kernel
+from torch import Size, Tensor
+from torch.nn import Module, Parameter
+
+
+class BasisExpansion(Module):
+    output_shape: Size
+    batch_shape: Optional[Size]
+
+
+class GeneralizedLinearBasis(BasisExpansion):
+    r"""Generalized linear basis functions:
+    `phi(x) = output_transform(input_transform(x)^T @ weight + bias)`."""
+
+    input_transform: Optional[TensorTransform]
+    output_transform: Optional[TensorTransform]
+
+    def __init__(
+        self,
+        initializer: GeneralizedLinearInitializer,
+        output_shape: Size,
+        batch_shape: Optional[Size] = None,
+    ) -> None:
+        super().__init__()
+        self.batch_shape = Size() if batch_shape is None else batch_shape
+        self.output_shape = output_shape
+        self.initializer = initializer
+        self._initialized = Parameter(
+            torch.zeros([], dtype=torch.bool), requires_grad=False
+        )
+
+    def forward(self, x: Tensor) -> Tensor:
+        if not self._initialized:
+            self.initialize(input_shape=x.shape[-1:])
+
+        x = x if self.input_transform is None else self.input_transform(x)
+        z = x @ self.weight.transpose(-2, -1)
```

Comment:
I'm not 100% certain, but it seems like there's a
`x = x.unsqueeze(1)`
needed here. In the MultiOutput / FullyBayesian case, we get:

`self.weight: num_outputs * 1 * num_dim * num_features`
`x: num_outputs *  num_queries * num_features`
'z: x @ self.weight.transpose(-2, -1) --> 
num_outputs * **num_outputs** * num_queries * num_features

With the proposed change, we get:
'z: x @ self.weight.transpose(-2, -1) --> 
num_outputs * **1** * num_queries * num_features

In the single output case, the functionality is unchanged.

Reply:
Hey! Thanks for pointing this out. I think there's a good chance you're right that there's something funny going here. Are you using the latest version of this PR (there was a bug in a previous version where `num_outputs` was not being passed)? If so I would sooner expect 
`self.weight: batch_shape x num_outputs x num_dim x num_features`,
which in many cases would resolve to
`self.weight: num_outputs x num_outputs x num_dim x num_features`.

Basically, my code (erroneously) views `model.batch_shape` and multi-output dimensions as separate things whereas BoTorch absorbs the multi-output dimension as part of the batch shape. Time allowing, I'll look into this and hopefully get back to you with a fix. If you happen to have a simple example that reproduces the issue that would be helpful.

Reply:
@hvarfner This issue should be resolved in the latest version of the PR. Note that `model.covar_module.batch_shape` now dictates the batch shape. Among other things, this means that we no longer maintain an explicit output dimension in the non-batch setting.